crates/lace-bin/src/typechecking/solver.rs

```diff
@@ -0,0 +1,197 @@
+use crate::lice::Lice;
+use crate::syntax_id::NodeId;
+use crate::typechecking::constraint::Constraint;
+use crate::typechecking::constraint_metadata::ConstraintMetadata;
+use crate::typechecking::context::{NameMapping, TypeContext, TypeMapping};
+use crate::typechecking::error::TypeError;
+use crate::typechecking::ty::{ConcreteType, Type, TypeVariable};
+use bumpalo::collections::Vec as BumpVec;
+use bumpalo::Bump;
+use unionfind::VecUnionFind;
+
+type SolveState = VecUnionFind<usize>;
+
+impl<'a> TypeContext<'a> {
+    #[allow(unused)]
+    fn cant_unify(&mut self, ca: ConcreteType, cb: ConcreteType, a: TypeVariable, b: TypeVariable) {
+        todo!()
+    }
+
+    fn unify_one(
+        &mut self,
+        a: TypeVariable,
+        b: TypeVariable,
+        meta: ConstraintMetadata<'a>,
+    ) -> TypeVariable {
+        // look up if any correspond to concrete types we know
+        let type_a = self.type_mapping.get(&a).copied();
+        let type_b = self.type_mapping.get(&b).copied();
+
+        match (type_a, type_b) {
+            // if not, we choose one by a random dice roll. My dice rolled 0 so we choose a
```

Comment:
You've got a weird die with a 0 value lol

Reply:
I have a binary die